File: src/components/alcohol-calc/UnitPicker.js
import React from 'react';
import { useTranslation } from 'react-i18next';
import { connect } from 'react-redux';
import { changeUnit } from '../../actions';

const isChecked = (unit, forUnit) => {
    return unit === forUnit;
}

const onChange = (props, unit) => {
    props.changeUnit(unit);
}

const UnitPicker = (props) => {
    const { t } = useTranslation();
    if (!props.unit) {
        return <div></div>;
    }
    return (
        <div className="grid grid-cols-12 gap-3 shadow-md p-4">
            <span className="2xl:col-span-3 xl:col-span-3 lg:col-span-4 md:col-span-5 sm:col-span-6 col-span-12 xs:col-span-12 2xl:text-right xl:text-right lg:text-right md:text-right sm:text-left xs:text-left text-left">{t('gravity unit')}</span>
            <div className="2xl:col-span-1 xl:col-span-1 lg:col-span-2 md:col-span-3 col-span-12 xs:col-span-12">
                <label>
                    <input type="radio" className="dark:bg-gray-700 dark:text-gray-300" id="brix" name="unit" checked={isChecked(props.unit, 'brix')} onChange={e => onChange(props, e.target.id)} />
                    &nbsp;&nbsp;{t('brix')}
                </label>
            </div>
            <div className="2xl:col-span-1 xl:col-span-1 lg:col-span-2 md:col-span-3 col-span-12 xs:col-span-12">
                <label>
                    <input type="radio" className="dark:bg-gray-700 dark:text-gray-300" id="plato" name="unit" checked={isChecked(props.unit, 'plato')} onChange={e => onChange(props, e.target.id)} />
                    &nbsp;&nbsp;{t('plato')}
                </label>
            </div>
        </div>
    );
}

const mapStateToProps = (state) => {
    return { unit: state.beerCalc.unit };
}

export default connect(mapStateToProps, { changeUnit })(UnitPicker);

File: src/components/unit_picker.rs
use crate::common::prelude::*;
use crate::STATE;
use dioxus::prelude::*;
use dioxus_i18n::t;

fn change_to_brix() {
    let original_gravity = STATE.read().original_gravity.to_brix();
    let final_gravity = STATE.read().final_gravity.to_brix();
    let mut new_state = STATE.read().clone();
    new_state.original_gravity = original_gravity;
    new_state.final_gravity = final_gravity;
    *STATE.write() = new_state;
}

fn change_to_plato() {
    let original_gravity = STATE.read().original_gravity.to_plato();
    let final_gravity = STATE.read().final_gravity.to_plato();
    let mut new_state = STATE.read().clone();
    new_state.original_gravity = original_gravity;
    new_state.final_gravity = final_gravity;
    *STATE.write() = new_state;
}

#[component]
pub fn GravityUnitPicker() -> Element {
    rsx! {
        div { class: "flex flex-row gap-4 shadow-md dark:shadow-neutral-600 items-baseline p-4",
            span { {format!("{}: ", t!("gravity_unit"))} }
            div {
                input {
                    r#type: "radio",
                    id: "brix",
                    class: "dark:bg-gray-700 dark:text-gray-300",
                    name: "gravity_unit",
                    checked: matches!(STATE.read().original_gravity, Gravity::Brix(_)),
                    onchange: move |_| change_to_brix(),
                }
                label { r#for: "brix", {format!(" {}", t!("brix"))} }
            }
            div {
                input {
                    r#type: "radio",
                    id: "plato",
                    class: "dark:bg-gray-700 dark:text-gray-300",
                    name: "gravity_unit",
                    checked: matches!(STATE.read().original_gravity, Gravity::Plato(_)),
                    onchange: move |_| change_to_plato(),
                }
                label { r#for: "plato", {format!(" {}", t!("plato"))} }
            }
        }
    }
}
